File: Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int start = 0, end = 0, max = 0;
        Set<Character> set = new HashSet<>();
        
        while(end < s.length()){
            if(!set.contains(s.charAt(end))){
                set.add(s.charAt(end++));
                max = Math.max(max, set.size());
            } else{
                set.remove(s.charAt(start++));
            }
        }
        return max;
    }
}

File: Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let i = 0, j = 0, max = 0;
    let set = new Set();
    
    while(j < s.length){
        if(!set.has(s[j])){
            set.add(s[j++]);
            max = Math.max(max, set.size);
        } else{
            set.delete(s[i++]);
        }
    }
    return max;
};
